starfish/image/_stack.py

```diff
@@ -1,54 +1,47 @@
-import json
 import os
 
 import numpy
+from slicedimage import Reader, Writer
 
-from ._base import ImageBase, ImageFormat
+from ._base import ImageBase
 
 
 class ImageStack(ImageBase):
-    def __init__(self, data, tile_format, num_hybs, num_chs, tile_shape):
-        self._data = data
-        self._tile_format = tile_format
+    def __init__(self, slicedimage):
+        self._slicedimage = slicedimage
+        self._num_hybs = slicedimage.get_dimension_shape('hyb')
+        self._num_chs = slicedimage.get_dimension_shape('ch')
+        self._tile_shape = tuple(slicedimage.default_tile_shape)
 
-        # shape data
-        self._num_hybs = num_hybs
-        self._num_chs = num_chs
-        self._tile_shape = tile_shape
+        self._data = numpy.zeros((self._num_hybs, self._num_chs) + self._tile_shape)
 
-        if len(tile_shape) == 2:
+        for tile in slicedimage.get_matching_tiles():
+            h = tile.indices['hyb']
+            c = tile.indices['ch']
+            self._data[h, c, :] = tile.numpy_array
+
+        if len(self._tile_shape) == 2:
             self._is_volume = False
         else:
             self._is_volume = True
 
     @classmethod
-    def from_org_json(cls, org_json_path):
-        with open(org_json_path, 'r') as in_file:
-            org_json = json.load(in_file)
-        metadata_dict = org_json['metadata']
-        num_hybs = metadata_dict['num_hybs']
-        num_chs = metadata_dict['num_chs']
-        tile_shape = tuple(metadata_dict['shape'])
-
-        data = numpy.zeros((num_hybs, num_chs) + tile_shape)
-        tile_format = ImageFormat[metadata_dict['format']]
-
-        data_dicts = org_json['data']
-        base_path = os.path.dirname(org_json_path)
+    def from_image_stack(cls, image_stack_name_or_url, baseurl):
+        slicedimage = Reader.parse_doc(image_stack_name_or_url, baseurl)
 
-        for data_dict in data_dicts:
-            h = data_dict['hyb']
-            c = data_dict['ch']
-            fname = data_dict['file']
-            im = tile_format.reader_func(os.path.join(base_path, fname))
-            data[h, c, :] = im
-
-        return ImageStack(data, tile_format, num_hybs, num_chs, tile_shape)
+        return ImageStack(slicedimage)
 
     @property
     def numpy_array(self):
         return self._data
 
+    @numpy_array.setter
+    def numpy_array(self, data):
+        for tile in self._slicedimage.get_matching_tiles():
```

Comment:
This name is also a bit confusing for me, since what we're doing is just getting all the tiles. `filter_tiles(filter=None)` is a bit clearer to me.

Related, since the setter doesn't directly set on the `self._data` attribute referenced in the property, I think this deserves a comment; it's almost certainly going to confuse contributors.

Reply:
Yeah, this part is not great IMO.  It might be that we need to treat `._data` as a cache of sorts, with dirty bits and all.

Let me ponder this.

Reply:
Sounds good. 👍